prov/rxd/src/rxd_ep.c

```diff
@@ -803,15 +805,27 @@ static int rxd_ep_close(struct fid *fid)
 	}
 
 	while (!dlist_empty(&ep->unexp_list)) {
-		dlist_pop_front(&ep->unexp_list, struct rxd_pkt_entry,
-				pkt_entry, d_entry);
-		ofi_buf_free(pkt_entry);
+		dlist_pop_front(&ep->unexp_list, struct rxd_unexp_msg,
+				unexp_msg, entry);
+		while (!dlist_empty(&unexp_msg->pkt_list)) {
+			dlist_pop_front(&unexp_msg->pkt_list, struct rxd_pkt_entry,
+					pkt_entry, d_entry);
+			ofi_buf_free(pkt_entry);
+		}
+		ofi_buf_free(unexp_msg->pkt_entry);
+		free(unexp_msg);
 	}
 
 	while (!dlist_empty(&ep->unexp_tag_list)) {
-		dlist_pop_front(&ep->unexp_tag_list, struct rxd_pkt_entry,
-				pkt_entry, d_entry);
-		ofi_buf_free(pkt_entry);
+		dlist_pop_front(&ep->unexp_tag_list, struct rxd_unexp_msg,
+				unexp_msg, entry);
+		while (!dlist_empty(&unexp_msg->pkt_list)) {
+			dlist_pop_front(&unexp_msg->pkt_list, struct rxd_pkt_entry,
+					pkt_entry, d_entry);
+			ofi_buf_free(pkt_entry);
+		}
+		ofi_buf_free(unexp_msg->pkt_entry);
+		free(unexp_msg);
```

Comment:
The code from the while statement down is duplicated twice here.  Try to isolate into function.  rxd_cleanup_unexp_msg()?